netket/stats.py

```diff
@@ -8,3 +8,30 @@ def subtract_mean(x):
     and over all MPI processes from each entry.
     """
     return _subtract_mean(x.reshape(-1, x.shape[-1]))
+
+
+from mpi4py import MPI
+import numpy as _np
+
+_MPI_comm = MPI.COMM_WORLD
+
+_n_nodes = _MPI_comm.Get_size()
+
+
+def mean(a, axis=None, dtype=None, out=None):
+    """
+    Compute the arithmetic mean along the specified axis and over MPI processes.
+
+    Returns the average of the array elements. The average is taken over the flattened array by default,
+    otherwise over the specified axis. float64 intermediate and return values are used for integer inputs.
+    """
+    me = _np.mean(a, axis=axis, dtype=None)
+
+    if out is None:
+        out = _np.copy(me)
+
+    _MPI_comm.Allreduce(me, out, op=MPI.SUM)
```

Comment:
Minor comment: Why are you copying in here? you can use MPI_IN_PLACE and perform the sum without allocating a copy. In C this would be done by calling it with the signature
`Allreduce(MPI_IN_PLACE, inout, op=MPI.SUM)`. Don't know if this works in mpi4py...

Reply:
I wouldn't know how to do that in mpi4py... the docs are a bit incomplete unfortunately !

Reply:
`Allreduce(MPI.IN_PLACE, me, op=MPI.SUM)` should work

Reply:
ah nope, there is an issue there 

  File "/Users/gcarleo/Works/NetKet/netket/netket/_vmc.py", line 223, in advance
    eloc -= _mean(eloc)
  File "/Users/gcarleo/Works/NetKet/netket/netket/stats.py", line 31, in mean
    _MPI_comm.Allreduce(MPI.IN_PLACE, out, op=MPI.SUM)
  File "mpi4py/MPI/Comm.pyx", line 713, in mpi4py.MPI.Comm.Allreduce
  File "mpi4py/MPI/msgbuffer.pxi", line 689, in mpi4py.MPI._p_msg_cco.for_allreduce
  File "mpi4py/MPI/msgbuffer.pxi", line 642, in mpi4py.MPI._p_msg_cco.for_cro_recv
  File "mpi4py/MPI/msgbuffer.pxi", line 154, in mpi4py.MPI.message_simple
  File "mpi4py/MPI/msgbuffer.pxi", line 93, in mpi4py.MPI.message_basic
  File "mpi4py/MPI/asbuffer.pxi", line 300, in mpi4py.MPI.getbuffer
  File "mpi4py/MPI/asbuffer.pxi", line 138, in mpi4py.MPI.PyMPI_GetBuffer
BufferError: scalar buffer is readonly

Reply:
I fixed it, for some reason Allreduce works only with arrays, not scalars... or maybe with scalars one has to explicitly pass the type